pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,296 @@
+}
+
+// HATrackerConfig contains the configuration require to
+// create a HA Tracker.
+type HATrackerConfig struct {
+	ReplicaLabel string
+	ClusterLabel string
+
+	KVStore ring.KVConfig
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.ReplicaLabel, "ha-tracker.replica", "replica", "Prometheus label to look for in samples to identify a Proemtheus HA replica.")
+	f.StringVar(&cfg.ClusterLabel, "ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Poemtheus HA cluster.")
+	// We want the ability to use different Consul instances for the ring and for HA cluster tracking.
+	cfg.KVStore.RegisterFlags(f, "ha-tracker")
+}
+
+// NewClusterTracker returns a new HA cluster tracker using either Consul
+// or in-memory KV store.
+func NewClusterTracker(cfg HATrackerConfig) (clusterTracker, error) {
+	codec := ring.ProtoCodec{Factory: ProtoReplicaDescFactory}
+
+	client, err := ring.NewKVStore(cfg.KVStore, codec)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t := &tracker{
+		logger:           util.Logger,
+		cfg:              cfg,
+		done:             make(chan struct{}),
+		elected:          map[string]replicaTracker{},
+		writeTimeout:     10 * time.Second,
+		overwriteTimeout: time.Minute,
+		client:           client,
+		cancel:           cancel,
+	}
+	go t.loop(ctx)
+	return t, nil
+}
+
+// Follows pattern used by ring for WatchKey.
+func (c *tracker) loop(ctx context.Context) {
+	defer close(c.done)
+	// The KVStore config we gave when creating c should have contained a prefix,
+	// which would have given us a prefixed KVStore client. So, we can pass empty string here.
+	c.client.WatchPrefix(ctx, "", func(value interface{}) bool {
+		if value == nil {
+			return true
+		}
+		v := value.(ring.KVEntry)
+		cluster := v.Cluster
+		replica := v.ReplicaEntry.(*ReplicaDesc)
+		c.setReplica(cluster, replica.GetReplica(), timestamp.Time(replica.GetTimestamp()))
+		return true
+	})
+}
+
+// Stop ends calls the trackers cancel function, which will end the loop for WatchPrefix.
+func (c *tracker) Stop() {
+	c.cancel()
+	<-c.done
+}
+
+// Wraps calls to client CAS with locking and error checking
+func (c *tracker) casWrapper(ctx context.Context, cluster, replica string, now time.Time) error {
+	// We have should a prefixed KVStore client, so we want to pass /cluster instead of just cluster.
+	// Prefixed client does prefix+key without separating the values.
+	err := c.client.CAS(ctx, fmt.Sprintf("/%s", cluster), func(in interface{}) (out interface{}, retry bool, err error) {
+		desc, ok := in.(*ReplicaDesc)
+		// There was either invalid or no data for the key, so we now accept samples
+		// from this replica.
+		if !ok || desc == nil {
+			desc = &ReplicaDesc{
+				Replica:   replica,
+				Timestamp: timestamp.FromTime(now),
+			}
+			return desc, false, nil
```

Comment:
Why do you not want to retry here?

Reply:
The only reason I can think _to retry_ is when there's an error talking to Consul. I supposed this could be one of those cases?